src/accountsdb/download.zig

```diff
@@ -0,0 +1,652 @@
+const Hash = lib.core.Hash;
+
+const DOWNLOAD_PROGRESS_UPDATES_NS = 30 * std.time.ns_per_s;
+
+const PeerSnapshotHash = struct {
+    contact_info: ContactInfo,
+    full_snapshot: SlotAndHash,
+    inc_snapshot: ?SlotAndHash,
+};
+
+const PeerSearchResult = struct {
+    is_me_count: usize = 0,
+    invalid_shred_version: usize = 0,
+    no_rpc_count: usize = 0,
+    no_snapshot_hashes_count: usize = 0,
+    is_blacklist: usize = 0,
+    is_valid: usize = 0,
+};
+
+/// finds valid contact infos which we can download a snapshot from.
+/// valid contact infos are:
+/// - not me
+/// - shred version matches
+/// - rpc socket is enabled
+/// - snapshot hash is available
+/// result is populated inside valid_peers (which is cleared at the beginning)
+pub fn findPeersToDownloadFromAssumeCapacity(
+    table: *const GossipTable,
+    contact_infos: []ContactInfo,
+    my_shred_version: usize,
+    my_pubkey: Pubkey,
+    blacklist: []Pubkey,
+    valid_peers: *std.ArrayList(PeerSnapshotHash),
+) PeerSearchResult {
+    valid_peers.clearRetainingCapacity();
+
+    var result = PeerSearchResult{};
+    search_loop: for (contact_infos) |*peer_contact_info| {
+        const is_me = peer_contact_info.pubkey.equals(&my_pubkey);
+        if (is_me) {
+            result.is_me_count += 1;
+            continue;
+        }
+
+        const matching_shred_version = my_shred_version == peer_contact_info.shred_version or my_shred_version == 0;
+        if (!matching_shred_version) {
+            result.invalid_shred_version += 1;
+            continue;
+        }
+        _ = peer_contact_info.getSocket(socket_tag.RPC) orelse {
+            result.no_rpc_count += 1;
+            continue;
+        };
+        const gossip_data = table.get(.{ .SnapshotHashes = peer_contact_info.pubkey }) orelse {
+            result.no_snapshot_hashes_count += 1;
+            continue;
+        };
+        const snapshot_hashes = gossip_data.value.data.SnapshotHashes;
+
+        var max_inc_hash: ?SlotAndHash = null;
+        for (snapshot_hashes.incremental) |inc_hash| {
+            if (max_inc_hash == null or inc_hash.slot > max_inc_hash.?.slot) {
+                max_inc_hash = inc_hash;
+            }
+        }
+
+        // dont try to download from a slow peer
+        for (blacklist) |black_list_peers| {
+            if (black_list_peers.equals(&peer_contact_info.pubkey)) {
+                result.is_blacklist += 1;
+                continue :search_loop;
+            }
+        }
+
+        valid_peers.appendAssumeCapacity(.{
+            // NOTE: maybe we need to deep clone here due to arraylist sockets?
+            .contact_info = peer_contact_info.*,
+            .full_snapshot = snapshot_hashes.full,
+            .inc_snapshot = max_inc_hash,
+        });
+    }
+    result.is_valid = valid_peers.items.len;
+
+    return result;
+}
+
+/// uses swap remove to remove peers that do not have a hash which is trusted
+pub fn removeUnTrustedSnapshotPeers(
```

Comment:
It probably doesn't matter, but my instinct would be to have this function directly return the trusted_snapshot_hashes, and then pass that into findPeersToDownloadFromAssumeCapacity, where it would be used as a whitelist. Then you wouldn't need to iterate over the list after the fact and remove items.